libevmasm/Assembly.cpp

```diff
@@ -348,6 +477,61 @@ Json::Value Assembly::assemblyJSON(map<string, unsigned> const& _sourceIndices)
 	return root;
 }
 
+bool Assembly::loadFromAssemblyJSON(std::string const& _source)
+{
+	solAssert(m_items.empty(), "");
+	solAssert(m_data.empty(), "");
+	solAssert(m_subs.empty(), "");
+	solAssert(m_auxiliaryData.empty(), "");
+	solAssert(m_usedTags == 1, "");
+
+	Json::Value assemblyJson;
+	bool success = util::jsonParseStrict(_source, assemblyJson);
+	if (success)
+	{
+		for (auto const& e: assemblyJson[".code"])
+		{
+			AssemblyItem item = assemblyItemFromJSON(e);
+			if (item.type() == AssemblyItemType::Tag)
+			{
+				auto tag = static_cast<unsigned>(item.data());
+				if (m_usedTags <= tag)
+					m_usedTags = tag + 1;
+			}
+			this->m_items.emplace_back(item);
+		}
+		Json::Value const& data = assemblyJson[".data"];
+		for(Json::ValueConstIterator itr = data.begin(); itr != data.end(); itr++)
+		{
+			solAssert(itr.key().isString(), "");
+			std::string key = itr.key().asString();
+			Json::Value const& code = data[key][".code"];
+			solAssert(code.isArray(), "");
+			Json::Value const& auxdata = data[key][".auxdata"];
+			if (auxdata.isString())
+			{
+				solAssert(m_auxiliaryData.empty(), "");
+				m_auxiliaryData = fromHex(auxdata.asString());
+			}
+			shared_ptr<Assembly> subassembly = make_shared<Assembly>();
+			for (auto const& e: code)
+			{
+				AssemblyItem item = assemblyItemFromJSON(e);
+				if (item.type() == AssemblyItemType::Tag)
+				{
+					auto tag = static_cast<unsigned>(item.data());
+					if (subassembly->m_usedTags <= tag)
+						subassembly->m_usedTags = tag + 1;
+				}
+				subassembly->m_items.emplace_back(item);
+			}
```

Comment:
This is pretty much the same as a few lines up, isn't it? Might be nice to unify...
Also doing sth like
```suggestion
			for (auto it = code.begin(); it != code.end(); ++it)
			{
			    auto const& e = *it;
				AssemblyItem item = assemblyItemFromJSON(e);
				if (item.type() == AssemblyItemType::Tag)
				{
					auto tag = static_cast<unsigned>(item.data());
					if (subassembly->m_usedTags <= tag)
						subassembly->m_usedTags = tag + 1;
					subassembly->m_items.emplace_back(item);
					if (std::next(it) != code->end() && *std::next(it) == AssemblyItem(Instruction::JumpDest)) // or however those are compared
					   ++it;
				}
				subassembly->m_items.emplace_back(item);
			}
```
would be *a* solution to the jumpdest issue, wouldn't it? Not sure it's the best one, though.

Reply:
Ah nice thx! I also ended up just ignoring all `JUMPDEST`, if they where coming from json. I'm not sure whether this is always correct. I guess I will know after implementing reimport tests for all semantic tests.

Reply:
That was mainly my point: if you just skip the ``JUMPDEST``s immediately following tags in this loop, then we can be sure that it's fine, since you'll still get any further stray ``JUMPDEST``s :-). (Although I could very well imagine that there are none anyways.)